far/vmenu.cpp

```diff
@@ -2129,253 +2331,139 @@ void VMenu::ShowMenu(bool IsParent)
 
-				if (!Items[I].Name.empty())
-				{
-					auto ItemWidth = static_cast<int>(Items[I].Name.size());
+		wchar_t PrevItem = (AnyPrev && Items[CurItemIndex - 1].Name.size() > I + PCorrection)? Items[CurItemIndex - 1].Name[I + PCorrection] : 0;
+		wchar_t NextItem = (AnyNext && Items[CurItemIndex + 1].Name.size() > I + NCorrection)? Items[CurItemIndex + 1].Name[I + NCorrection] : 0;
 
-					if (ItemWidth + 2 > m_Where.width())
-						ItemWidth = m_Where.width() - 2;
+		if (!PrevItem && !NextItem)
+			break;
 
-					GotoXY(m_Where.left + (m_Where.width() - 2 - ItemWidth) / 2, Y);
-					Text(concat(L' ', fit_to_left(Items[I].Name, ItemWidth), L' '));
-				}
-			}
+		if (PrevItem == BoxSymbols[BS_V1])
+		{
+			if (NextItem == BoxSymbols[BS_V1])
+				separator[I + (BoxType == NO_BOX?1:2) + 1] = BoxSymbols[BS_C_H1V1];
 			else
-			{
-				if (m_BoxType!=NO_BOX)
-				{
-					SetColor(Colors[VMenuColorBox]);
-					Text(BoxChar);
-					GotoXY(m_Where.right, Y);
-					Text(BoxChar);
-				}
-
-				GotoXY(m_Where.left + (m_BoxType == NO_BOX? 0 : 1), Y);
-
-				FarColor CurColor;
-				if ((Items[I].Flags&LIF_SELECTED))
-					CurColor = Colors[Items[I].Flags & LIF_GRAYED? VMenuColorSelGrayed : VMenuColorSelected];
-				else
-					CurColor = Colors[Items[I].Flags & LIF_DISABLE? VMenuColorDisabled : (Items[I].Flags & LIF_GRAYED? VMenuColorGrayed : VMenuColorText)];
-
-				SetColor(CurColor);
-
-				string strMenuLine;
-				wchar_t CheckMark = L' ';
-
-				if (Items[I].Flags & LIF_CHECKED)
-				{
-					if (!(Items[I].Flags & 0x0000FFFF))
-						CheckMark = 0x221A;
-					else
-						CheckMark = static_cast<wchar_t>(Items[I].Flags & 0x0000FFFF);
-				}
-
-				strMenuLine.push_back(CheckMark);
-				strMenuLine.push_back(L' '); // left scroller (<<) placeholder
-				const auto PrefixSize = strMenuLine.size();
-
-				size_t HotkeyVisualPos = string::npos;
-				auto MenuItemForDisplay = CheckFlags(VMENU_SHOWAMPERSAND)? Items[I].Name : HiText2Str(Items[I].Name, &HotkeyVisualPos);
+				separator[I + (BoxType == NO_BOX?1:2) + 1] = BoxSymbols[BS_B_H1V1];
+		}
+		else if (NextItem == BoxSymbols[BS_V1])
+		{
+			separator[I + (BoxType == NO_BOX?1:2) + 1] = BoxSymbols[BS_T_H1V1];
+		}
+	}
+}
 
-				MenuItemForDisplay.erase(0, Items[I].ShowPos);
+void VMenu::ApplySeparatorName(const MenuItemEx& CurItem, string& separator) const
+{
+	if (CurItem.Name.empty() || separator.size() <= 3)
+		return;
 
-				if (HotkeyVisualPos != string::npos)
-				{
-					if (HotkeyVisualPos < Items[I].ShowPos)
-						HotkeyVisualPos = string::npos;
-					else
-						HotkeyVisualPos -= Items[I].ShowPos;
-				}
+	auto NameWidth{ std::min(CurItem.Name.size(), separator.size() - 2) };
+	auto NamePos{ (separator.size() - NameWidth) / 2 };
 
-				// fit menu string into available space
-				bool ShowRightScroller = false;
-				if (MenuItemForDisplay.size() > MaxLineWidth)
-				{
-					MenuItemForDisplay.resize(MaxLineWidth);
-					ShowRightScroller = true;
-				}
+	separator.at(NamePos - 1) = L' ';
```

Comment:
Why `at`, do we expect exceptions?

Reply:
Preexisting code, but of course an exception is not expected. Fixed.

Reply:
If you're about `VMenu::at` - that's different.
VMenu is usually accessed via a (smart) pointer and writing `menu->at(idx)` is way more natural than `(*menu)[idx]`.

Reply:
Ah, sorry, it was I who introduced it here. I actually merged my changed from a different branch where I could not be so sure that the `separator` string is of correct size. It should not happen in this context.